prov/util/src/util_cntr.c

```diff
@@ -0,0 +1,335 @@
+ * SOFTWARE.
+ */
+
+#include <stdlib.h>
+#include <string.h>
+
+#include <fi_enosys.h>
+#include <fi_util.h>
+
+int ofi_check_cntr_attr(const struct fi_provider *prov,
+			const struct fi_cntr_attr *attr)
+{
+	if (!attr)
+		return FI_SUCCESS;
+
+        if (attr->flags) {
+		FI_WARN(prov, FI_LOG_CQ, "unsupported flags\n");
+		return -FI_EINVAL;
+	}
+
+	switch (attr->wait_obj) {
+	case FI_WAIT_NONE:
+		break;
+	case FI_WAIT_SET:
+		if (!attr->wait_set) {
+			FI_WARN(prov, FI_LOG_CNTR, "invalid wait set\n");
+			return -FI_EINVAL;
+		}
+		/* fall through */
+	case FI_WAIT_UNSPEC:
+	case FI_WAIT_FD:
+		break;
+	default:
+		FI_WARN(prov, FI_LOG_CNTR, "unsupported wait object\n");
+		return -FI_EINVAL;
+	}
+
+	return 0;
+}
+
+static uint64_t ofi_cntr_read(struct fid_cntr *cntr_fid)
+{
+	struct util_cntr *cntr = container_of(cntr_fid, struct util_cntr, cntr_fid);
+	assert(cntr->cntr_fid.fid.fclass == FI_CLASS_CNTR);
+
+	cntr->progress(cntr);
+
+	return ofi_atomic_get64(&cntr->cnt);
+}
+
+static uint64_t ofi_cntr_readerr(struct fid_cntr *cntr_fid)
+{
+	struct util_cntr *cntr = container_of(cntr_fid, struct util_cntr, cntr_fid);
+	assert(cntr->cntr_fid.fid.fclass == FI_CLASS_CNTR);
+
+	cntr->progress(cntr);
+
+	return ofi_atomic_get64(&cntr->err);
+}
+
+static int ofi_cntr_add(struct fid_cntr *cntr_fid, uint64_t value)
+{
+	struct util_cntr *cntr = container_of(cntr_fid, struct util_cntr, cntr_fid);
+	assert(cntr->cntr_fid.fid.fclass == FI_CLASS_CNTR);
+
+	ofi_atomic_add64(&cntr->cnt, value);
+
+	if(cntr->wait)
+		cntr->wait->signal(cntr->wait);
+
+	return FI_SUCCESS;
+}
+
+static int ofi_cntr_adderr(struct fid_cntr *cntr_fid, uint64_t value)
+{
+	struct util_cntr *cntr = container_of(cntr_fid, struct util_cntr, cntr_fid);
+	assert(cntr->cntr_fid.fid.fclass == FI_CLASS_CNTR);
+
+	ofi_atomic_add64(&cntr->err, value);
+
+	if(cntr->wait)
+		cntr->wait->signal(cntr->wait);
+
+	return FI_SUCCESS;
+}
+
+static int ofi_cntr_set(struct fid_cntr *cntr_fid, uint64_t value)
+{
+	struct util_cntr *cntr = container_of(cntr_fid, struct util_cntr, cntr_fid);
+	assert(cntr->cntr_fid.fid.fclass == FI_CLASS_CNTR);
+
+	ofi_atomic_initialize64(&cntr->cnt, value);
+
+	if(cntr->wait)
+		cntr->wait->signal(cntr->wait);
+
+	return FI_SUCCESS;
+}
+
+static int ofi_cntr_seterr(struct fid_cntr *cntr_fid, uint64_t value)
+{
+	struct util_cntr *cntr = container_of(cntr_fid, struct util_cntr, cntr_fid);
+	assert(cntr->cntr_fid.fid.fclass == FI_CLASS_CNTR);
+
+	ofi_atomic_initialize64(&cntr->err, value);
+
+	if(cntr->wait)
+		cntr->wait->signal(cntr->wait);
+
+	return FI_SUCCESS;
+}
+
+static int ofi_cntr_wait(struct fid_cntr *cntr_fid, uint64_t threshold, int timeout)
+{
+	struct util_cntr *cntr = container_of(cntr_fid, struct util_cntr, cntr_fid);
+	uint64_t current_ms;
+	uint64_t finish_ms;
+	uint64_t err = ofi_cntr_readerr(cntr_fid);
+
+	assert(cntr->cntr_fid.fid.fclass == FI_CLASS_CNTR);
+
+	if (threshold <= ofi_cntr_read(cntr_fid))
```

Comment:
I understand cntr_read and cntr_readerr call progress.  My intent was to only call progress once and read the counters directly using the atomic calls, so that we avoid the overhead of multiple progress calls.  For example, at this point in the code, we've tried to progress the counter's endpoints twice -- once from ofi_cntr_readerr, and again with ofi_cntr_read.